### PR TITLE
perf(flask): small enhancements

### PR DESCRIPTION
**Pull request by P403n1x87**
## Description

The performance of the flask integration has been enhanced slightly by caching some string operations in the middleware wrappers and by some other small refactoring.

## Checklist
- [ ] Added to the correct milestone.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

File: ddtrace/contrib/flask/middleware.py [hunk 1]
```diff
@@ -20,6 +21,11 @@
 SPAN_NAME = "flask.request"
 
 
+@cached()
+def _normalize_resource(resource):
+    return compat.to_unicode(resource).lower()
+
+
 class TraceMiddleware(object):
     @deprecated(message="Use patching instead (see the docs).", version="1.0.0")
```
Comment (brettlangdon) [inline ddtrace/contrib/flask/middleware.py:30]: you probably don't need to update this middleware, it is deprecated and is unlikely that it is used by anyone 🤞🏻

File: ddtrace/contrib/flask/patch.py [hunk 2]
```diff
@@ -274,54 +299,27 @@ def traced_wsgi_app(pin, wrapped, instance, args, kwargs):
...
-
-                trace_utils.set_http_meta(s, config.flask, status_code=code, response_headers=headers)
-                return func(status_code, headers)
-
-            return traced_start_response
-
-        start_response = _wrap_start_response(start_response)
+            span.set_tag(ANALYTICS_SAMPLE_RATE_KEY, sample_rate)
+
+        span.set_tag(FLASK_VERSION, flask_version_str)
```
Comment (brettlangdon) [inline ddtrace/contrib/flask/patch.py:315]: `_set_tag_str`?

File: ddtrace/contrib/flask/patch.py [hunk 3]
```diff
@@ -472,20 +470,20 @@ def _traced_request(pin, wrapped, instance, args, kwargs):
 
             # DEV: This name will include the blueprint name as well (e.g. `bp.index`)
             if not span.get_tag(FLASK_ENDPOINT) and request.endpoint:
-                span.resource = u"{} {}".format(request.method, request.endpoint)
+                span.resource = u" ".join((request.method, request.endpoint))
                 span.set_tag(FLASK_ENDPOINT, request.endpoint)
```
Comment (brettlangdon) [inline ddtrace/contrib/flask/patch.py:474]: `_set_tag_str` .... ?

I think this one should be a string...

File: ddtrace/contrib/flask/middleware.py (hunk 1 above)
Comment (P403n1x87) [reply ddtrace/contrib/flask/middleware.py:30]: yeah I wasn't sure but decided to touch it anyway, just in case not everybody has migrated away.

**Review by Kyle-Verhoog — COMMENTED** [review]
Just one concern otherwise lgtm!

File: ddtrace/contrib/flask/middleware.py [hunk 4]
```diff
@@ -96,9 +102,8 @@ def _teardown_request(self, exception):
         will be passed in.
         """
         # when we teardown the span, ensure we have a clean slate.
-        span = getattr(g, "flask_datadog_span", None)
-        setattr(g, "flask_datadog_span", None)
-        if not span:
+        span = g.__dict__.pop("flask_datadog_span", None)
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/contrib/flask/middleware.py:105]: If `g` were a `__slots__` object, wouldn't the original `setattr` fail? 🤔

[the remark this answers is not in the document]
Comment (Kyle-Verhoog) [reply ddtrace/contrib/flask/middleware.py:105]: ah fair enough! I guess we do make that assumption a lot that objects are not using slots 😩

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by nizox — APPROVED** [review]

**Pull request merged**